src/python/pants/engine/scheduler.py

```diff
@@ -122,86 +120,75 @@ def _to_ids_buf(self, types):
-                                                        self._to_utf8_buf(field),
-                                                        self._to_constraint(selector.input_product))
-          else:
-            raise ValueError('Unrecognized Selector type: {}'.format(selector))
-        self._native.lib.task_end(self._scheduler)
+        if type(rule) is SingletonRule:
+          self._register_singleton(output_constraint, rule)
+        elif type(rule) is TaskRule:
+          self._register_task(output_constraint, rule)
+        else:
+          raise ValueError('Unexpected Rule type: {}'.format(rule))
+
+  def _register_singleton(self, output_constraint, rule):
+    """Register the given SingletonRule.
+
+    Singleton tasks are those that are the default for a particular type(product).
+    """
+    self._native.lib.singleton_add(self._scheduler,
+                                   self._to_value(rule.value),
+                                   output_constraint)
+
+  def _register_task(self, output_constraint, rule):
+    """Register the given TaskRule with the native scheduler."""
+    input_selects = rule.input_selectors
+    func = rule.func
+    self._native.lib.task_add(self._scheduler, Function(self._to_id(func)), output_constraint)
+    for selector in input_selects:
+      selector_type = type(selector)
+      product_constraint = self._to_constraint(selector.product)
+      if selector_type is Select:
+        self._native.lib.task_add_select(self._scheduler,
+                                          product_constraint)
+      elif selector_type is SelectVariant:
+        key_buf = self._to_utf8_buf(selector.variant_key)
+        self._native.lib.task_add_select_variant(self._scheduler,
+                                                  product_constraint,
+                                                  key_buf)
+      elif selector_type is SelectDependencies:
+        self._native.lib.task_add_select_dependencies(self._scheduler,
+                                                      product_constraint,
+                                                      self._to_constraint(selector.dep_product),
+                                                      self._to_utf8_buf(selector.field),
+                                                      self._to_ids_buf(selector.field_types))
+      elif selector_type is SelectTransitive:
+        self._native.lib.task_add_select_transitive(self._scheduler,
+                                                    product_constraint,
+                                                    self._to_constraint(selector.dep_product),
+                                                    self._to_utf8_buf(selector.field),
+                                                    self._to_ids_buf(selector.field_types))
+      elif selector_type is SelectProjection:
+        if len(selector.fields) != 1:
+          raise ValueError("TODO: remove support for projecting multiple fields at once.")
```

Comment:
Maybe: `Projecting multiple fields at once is not supported.` I think this should fail when constructing the selector, so that the error is more actionable.